File: prob_60.py
#leetcode 994
#have a dir array and accordingly access all the elements and process them with simultaneously decrementing the fresh oranges count and incrementing minutes.

class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        visited=[[False for _ in range(len(grid[0]))] for _ in range(len(grid))]
        minutes=-1
        fresh=0
        q=[]

        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j]==2:
                    q.append((i,j))
                    visited[i][j]=True
                elif grid[i][j]==1:
                    fresh=fresh+1

        if fresh==0: return 0

        dir=[(0,1),(1,0),(-1,0),(0,-1)]
        while q:
            size=len(q)
            for i in range(size):
                q_con=q[i]
                r,c=q_con
                for (x,y) in dir:
                    x_new,y_new=r+x,c+y
                    if ( (0<=x_new<len(grid)) and (0<=y_new<len(grid[0])) ):
                        if not visited[x_new][y_new] and grid[x_new][y_new]==1:
                            q.append((x_new,y_new))
                            grid[x_new][y_new]=2
                            visited[x_new][y_new]=True
                            fresh -= 1
            minutes=minutes+1 #Once the level is done we are increasing the minutes by 1
            q=q[size:]
        if (fresh==0):
            return minutes
        else:
            -1

File: prob60.java
// S30 Big N Problem #60 {Medium}
// 994. Rotting Oranges
// Time Complexity :O(m*n) where m and n are the number of rows and columns of the matrix
// Space Complexity :O(m*n) where m and n are number of rows and columns of the matrix (stored in queue)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :None

// Your code here along with comments explaining your approach
// BFS solution
class Solution {
    public int orangesRotting(int[][] grid) {
        int m=grid.length;
        if(grid==null || m ==0) return 0;
        int n=grid[0].length;
        int fresh=0;
        Queue <int []> q=new LinkedList<>();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==2) q.add(new int[] {i,j});
                if(grid[i][j]==1) fresh++;
            }
        }
        
        int minutes=0;
        int [][]dirs={{0,1},{0,-1},{-1,0},{1,0}};
        while(!q.isEmpty()){
            int size=q.size();
            for(int i=0;i<size;i++){
                int [] rotten=q.poll();
                for(int []dir:dirs){
                    int r=rotten[0]+dir[0];
                    int c=rotten[1]+dir[1];
                    if(r>=0 && r<m && c>=0 && c<n && grid[r][c]==1){
                        grid[r][c]=2;
                        q.add(new int[]{r,c});
                        fresh--;
                    }
                }
            }
            minutes++;
        }
        if(fresh>0) return -1;
        return minutes>0 ? minutes-1:0; //minutes-1 because for first level - oranges are already rotten 
    }
}